packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,255 @@
+          void onClose(connection, reason);
+        }
+      },
+      async onReceive(connection, bytes) {
+        const { onReceive } = this.state;
+        if (onReceive && typeof onReceive === 'function') {
+          void onReceive(connection, bytes);
+        }
+        return '';
+      },
+    },
+  );
+
+export const OrchestrationI = M.interface('Orchestration', {
+  provideAccount: M.callWhen(M.string(), M.string())
+    .optional(M.remotable('Port'))
+    .returns(M.remotable('ChainAccount')),
+  getChain: M.callWhen(M.string()).returns(M.remotable('Chain')),
+});
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<typeof prepareChainAccount>} createChainAccount
+ * @param {ReturnType<typeof prepareConnectionHandler>} makeConnectionHandler
+ */
+const prepareOrchestration = (
+  zone,
+  createChainAccount,
+  makeConnectionHandler,
+) =>
+  zone.exoClassKit(
+    'Orchestration',
+    {
+      admin: M.interface('OrchestrationAdmin', {
+        foo: M.call().returns(M.string()),
+      }),
+      public: OrchestrationI,
+    },
+    /** @param {Partial<OrchestrationPowers>} [initialPowers] */
+    initialPowers => {
+      /** @type {PowerStore} */
+      const powers = zone.detached().mapStore('PowerStore');
+      if (initialPowers) {
+        for (const [name, power] of Object.entries(initialPowers)) {
+          powers.init(/** @type {keyof OrchestrationPowers} */ (name), power);
+        }
+      }
+      return { powers, icaControllerNonce: 0 };
+    },
+    {
+      admin: {
+        foo() {
+          return 'bar';
+        },
+      },
+      public: {
+        /**
+         * @param {string} hostConnectionId the counterparty connection_id
+         * @param {string} controllerConnectionId self connection_id
+         * @param {Port} [currentPort] if a port is provided, it will be reused
+         *   to create the account
+         */
+        async provideAccount(
+          hostConnectionId,
+          controllerConnectionId,
+          currentPort,
+        ) {
+          await null;
+          const connectionHandler = makeConnectionHandler();
+          let port;
+          if (!matches(currentPort, M.remotable('Port'))) {
+            if (currentPort) trace('Invalid Port provided, binding a new one.');
+            const network = getPower(this.state.powers, 'network');
+
+            port = await E(network)
+              .bind(`/ibc-port/icacontroller-${this.state.icaControllerNonce}`)
+              .catch(e => Fail`Failed to bind port: ${bare(e)}`);
+
+            this.state.icaControllerNonce += 1;
+          } else {
+            port = currentPort;
+          }
+          const connString = JSON.stringify({
+            version: 'ics27-1',
+            controllerConnectionId, // self_connection_id
+            hostConnectionId, // counterparty_connection_id
+            address: '', // will be filled in by the counterparty
+            encoding: 'proto3',
```

Comment:
To do: see if chains will successfully negotiate `proto3json`

Reply:
Attempted `json` and `proto3-json`. _Could not find an enum for `version` in any spec/reference materials, if someone has a pointer._

> unsupported encoding format **proto3-json**: codec is not supported [cosmos/ibc-go/v7@v7.3.1/modules/apps/27-interchain-accounts/types/metadata.go:104]

> unsupported encoding format **json**: codec is not supported [cosmos/ibc-go/v7@v7.3.1/modules/apps/27-interchain-accounts/types/metadata.go:104]